worker/lease/manager.go

```diff
@@ -520,27 +519,30 @@ func (manager *Manager) checkBlocks(blocks blocks) {
 
 // computeNextTimeout iterates the leases and finds out what the next time we
 // want to wake up, expire any leases and then handle any unblocks that happen.
-// It is based on the MaxSleep time, and on any expire that is going to expire
-// after now but before MaxSleep.
-// it's worth checking for stalled collaborators.
+// It is the earliest lease expiration due in the future, but before MaxSleep.
 func (manager *Manager) computeNextTimeout(lastTick time.Time, leases map[lease.Key]lease.Info) {
 	now := manager.config.Clock.Now()
 	nextTick := now.Add(manager.config.MaxSleep)
 	for _, info := range leases {
 		if !info.Expiry.After(lastTick) {
-			// The previous expire will expire this lease eventually,
-			// or the manager will die with an error.
-			// Either way, we don't need to worry about expirations for
-			// a previous tick here.
+			// We expect the lease clock tick processing to expire this lease
+			// soon. It is not our responsibility to handle this case.
+			manager.config.Logger.Tracef("[%s] found lease holder past expiry: %s", manager.logContext, info.Holder)
 			continue
 		}
 		if info.Expiry.After(nextTick) {
 			continue
 		}
 		nextTick = info.Expiry
 	}
-	manager.config.Logger.Tracef("[%s] next expire in %v %v",
-		manager.logContext, nextTick.Sub(now).Round(time.Millisecond), nextTick)
+
+	// The lease clock is ticked every second, so sub-second durations
+	// can cause block checks resulting in no action. This is wasteful because
+	// accessing leases from the FSM causes read-locks, potentially
+	// contributing to contention. Round *up* to the nearest second.
+	nextDuration := (nextTick.Sub(now) + 500*time.Millisecond).Round(time.Second)
```

Comment:
I wonder about finding the min, and just always adding 1s to give time for things to batch a bit more. It won't really be problematic and might allow a little bit more batching.

Given that the only thing that actually expires the leases is coordination with the FSM there will always be a bit of estimation error. eg we think the lease will expire in 950ms but the clock won't actually tick for at least 1s, and if we wake up right at 1s, but it ticks (plus FSM sync time) it doesn't wake up until 1.1s and then we miss, and end up hitting the "next expire before now" case. And if you hit that acase 

I wonder if we would actually want:

max(now, min(all_expiries_even_in_the_past)) + 1s

Thoughts?

Reply:
This is done.